File: web/js/report.js
// Report screen: on-page preview + Download PDF link (Phase 5).
// PDF bytes land in Phase 6; the link targets that endpoint already.
(function () {
  const $ = (id) => document.getElementById(id);
  const fmt = (n, d) =>
    n === null || n === undefined || !Number.isFinite(Number(n))
      ? "n/a"
      : Number(n).toFixed(d === undefined ? 1 : d);

  function pair(k, v) {
    return `<tr><th scope="row">${k}</th><td>${v}</td></tr>`;
  }

  function idFromQuery() {
    const q = new URLSearchParams(window.location.search).get("id");
    return q && q.trim() ? q.trim() : "";
  }

  function fmtTime(t) {
    if (!t) return "—";
    try {
      return new Date(t).toLocaleString();
    } catch {
      return String(t);
    }
  }

  async function load(id) {
    const note = $("report-note");
    if (!id) {
      if (note) note.textContent = "Enter an experiment ID first.";
      return;
    }
    if (note) note.textContent = `Loading ${id}…`;
    try {
      const [exp, m] = await Promise.all([
        window.DDoSLabAPI.getExperiment(id),
        window.DDoSLabAPI.getMetrics(id).catch(() => null),
      ]);
      const cfg = exp.config || {};
      const defense = cfg.defense ? "on — rate limiting" : "off — baseline";
      $("sum-rows").innerHTML =
        pair("Experiment", `<code>${exp.id}</code>`) +
        pair("Status", exp.status) +
        pair("Endpoint", `${exp.target_url || cfg.endpoint || "—"}`) +
        pair("Started", fmtTime(exp.started_at)) +
        pair("Completed", fmtTime(exp.completed_at)) +
        pair("Config", `${cfg.duration_seconds ?? "—"}s · ${cfg.requests_per_second ?? "—"} RPS · ${cfg.workers ?? "—"} workers · defense ${defense}`);

      const total = m ? m.total_requests : exp.total_requests;
      const ok = m ? m.successful_requests : exp.successful_requests;
      const fail = m ? m.failed_requests : exp.failed_requests;
      const errRate = total > 0 ? ((fail / total) * 100).toFixed(1) + "%" : "n/a";
      $("traffic-rows").innerHTML =
        pair("Total requests", total ?? "—") +
        pair("Successful", ok ?? "—") +
        pair("Failed", fail ?? "—") +
        pair("Error rate", errRate) +
        pair("RPS", m ? fmt(m.rps) : "n/a (run finished; live RPS unavailable)");

      $("latency-rows").innerHTML = m
        ? pair("Average", `${fmt(m.avg_latency_ms)} ms`) +
          pair("P50", `${fmt(m.p50_latency_ms)} ms`) +
          pair("P95", `${fmt(m.p95_latency_ms)} ms`) +
          pair("P99", `${fmt(m.p99_latency_ms)} ms`)
        : `<tr><td>Metrics unavailable for this experiment.</td></tr>`;

      const codes = m ? m.status_codes || {} : {};
      const codeRows = Object.keys(codes).length
        ? Object.keys(codes)
            .sort()
            .map((c) => pair(`HTTP ${c}`, codes[c]))
            .join("")
        : pair("Status codes", "none recorded");
      const sys = m
        ? pair(
            "Target CPU / RSS",
            m.cpu_percent === null && m.memory_rss_bytes === null
              ? "n/a (non-Linux target)"
              : `${fmt(m.cpu_percent)}% / ${m.memory_rss_bytes === null ? "n/a" : (m.memory_rss_bytes / 1048576).toFixed(1) + " MiB"}`
          )
        : "";
      $("error-rows").innerHTML = codeRows + sys;

      const pdf = $("pdf-link");
      pdf.href = window.DDoSLabAPI.reportPdfUrl(id);
      pdf.setAttribute("download", `ddoslab-experiment-${id}.pdf`);
      const back = $("back-live");
      if (back) back.href = `/live.html?id=${encodeURIComponent(id)}`;
      if (note) note.textContent = `Showing ${id} (${exp.status}).`;
    } catch (err) {
      if (note) note.textContent = `error: ${err.message}`;
    }
  }

  function shortId(id) {
    return id && id.length > 8 ? id.slice(0, 8) : id || "?";
  }

  // Compare two runs side by side (UI only; comparison PDF is V2).
  async function compare(idA, idB) {
    const note = $("compare-note");
    const table = $("compare-table");
    if (!idA || !idB) {
      if (note) note.textContent = "Enter both experiment IDs first.";
      return;
    }
    if (note) note.textContent = "Loading…";
    try {
      const [a, b] = await Promise.all([
        Promise.all([window.DDoSLabAPI.getExperiment(idA), window.DDoSLabAPI.getMetrics(idA).catch(() => null)]),
        Promise.all([window.DDoSLabAPI.getExperiment(idB), window.DDoSLabAPI.getMetrics(idB).catch(() => null)]),
      ]);
      const rows = [
        ["Defense", (a[0].config || {}).defense ? "on" : "off", (b[0].config || {}).defense ? "on" : "off"],
        ["Status", a[0].status, b[0].status],
        ["Total requests", total(a[1], a[0]), total(b[1], b[0])],
        ["Error rate", rate(a[1], a[0]), rate(b[1], b[0])],
        ["P50 (ms)", lat(a[1], "p50_latency_ms"), lat(b[1], "p50_latency_ms")],
        ["P95 (ms)", lat(a[1], "p95_latency_ms"), lat(b[1], "p95_latency_ms")],
        ["P99 (ms)", lat(a[1], "p99_latency_ms"), lat(b[1], "p99_latency_ms")],
        ["HTTP 429", code(a[1], "429"), code(b[1], "429")],
        ["HTTP 5xx", fivexx(a[1]), fivexx(b[1])],
      ];
      $("cmp-a-h").textContent = `A (${shortId(idA)})`;
      $("cmp-b-h").textContent = `B (${shortId(idB)})`;
      $("compare-rows").innerHTML = rows
        .map(([k, x, y]) => `<tr><th scope="row">${k}</th><td>${x}</td><td>${y}</td></tr>`)
        .join("");
      table.hidden = false;
      if (note) note.textContent = "Measured values only — decide for yourself which trade-off is better.";
    } catch (err) {
      if (note) note.textContent = `error: ${err.message}`;
    }
  }

  function total(m, exp) {
    if (m) return String(m.total_requests);
    return String(exp.total_requests ?? "n/a");
  }

  function rate(m, exp) {
    const total = m ? m.total_requests : exp.total_requests;
    const fail = m ? m.failed_requests : exp.failed_requests;
    if (!total) return "n/a";
    return `${((fail / total) * 100).toFixed(1)}%`;
  }

  function lat(m, k) {
    if (!m || m[k] === null || m[k] === undefined) return "n/a";
    return Number(m[k]).toFixed(1);
  }

  function code(m, c) {
    if (!m || !m.status_codes) return "n/a";
    return String(m.status_codes[c] || 0);
  }

  function fivexx(m) {
    if (!m || !m.status_codes) return "n/a";
    return String(
      Object.entries(m.status_codes)
        .filter(([k]) => k.startsWith("5"))
        .reduce((n, [, v]) => n + v, 0)
    );
  }

  function init() {
    const params = new URLSearchParams(window.location.search);
    const qid = (params.get("id") || "").trim();
    const qcmp = (params.get("compare") || "").trim();
    if (qid) $("report-id").value = qid;
    if (qcmp) $("compare-id").value = qcmp;
    $("report-load")?.addEventListener("click", () => load($("report-id").value.trim()));
    $("compare-load")?.addEventListener("click", () =>
      compare($("report-id").value.trim(), $("compare-id").value.trim())
    );
    if (qid) load(qid);
    if (qid && qcmp) compare(qid, qcmp);
  }

  if (document.readyState === "loading") {
    document.addEventListener("DOMContentLoaded", init);
  } else {
    init();
  }
})();

File: backend/internal/report/report.go
// Package report builds the Phase 6 PDF experiment report.
//
// Pure-Go rendering via go-pdf/fpdf (the maintained gofpdf continuation):
// no headless browser, A4 portrait, generous margins, one accent color,
// labeled value pairs over dense charts. See context/architecture/01-pdf-reports.md.
package report

import (
	"bytes"
	"fmt"
	"sort"
	"time"

	"github.com/go-pdf/fpdf"
)

// Input is the report data contract. The experiment handler maps its Result
// onto this DTO so this package never imports the experiment package
// (which would be an import cycle via the handler).
type Input struct {
	ID                 string
	Status             string
	StartedAt          time.Time
	CompletedAt        *time.Time // nil while running
	TargetURL          string
	Defense            bool
	DurationSeconds    int
	RequestsPerSecond  int
	Workers            int
	TotalRequests      int64
	SuccessfulRequests int64
	FailedRequests     int64
	RPS                float64
	AvgLatencyMs       float64
	P50LatencyMs       float64
	P95LatencyMs       float64
	P99LatencyMs       float64
	Samples            int
	StatusCodes        map[string]int64
	CPUPercent         *float64
	MemoryRSSBytes     *int64
}

const (
	accentR, accentG, accentB = 37, 99, 235 // restrained blue
	marginMM                 = 18.0
)

// Build renders res as a complete PDF document.
func Build(res Input) ([]byte, error) {
	pdf := fpdf.New("P", "mm", "A4", "")
	pdf.SetMargins(marginMM, 18, marginMM)
	pdf.SetAutoPageBreak(true, 20)
	pdf.SetTitle(fmt.Sprintf("DDoSLab Experiment Report — %s", res.ID), false)
	pdf.SetAuthor("DDoSLab", false)

	pdf.AddPage()
	footer(pdf)

	coverStrip(pdf, res)
	section(pdf, "Traffic configuration")
	pairs(pdf, configRows(res))
	rule(pdf)

	section(pdf, "Traffic")
	pairs(pdf, trafficRows(res))
	rule(pdf)

	section(pdf, "Performance")
	pairs(pdf, latencyRows(res))
	rule(pdf)

	section(pdf, "Server impact")
	pairs(pdf, serverRows(res))
	rule(pdf)

	section(pdf, "Observations")
	for _, o := range Observe(res) {
		pdf.SetFont("Helvetica", "", 10)
		pdf.SetTextColor(30, 30, 30)
		pdf.MultiCell(0, 5.5, "•  "+o, "", "L", false)
		pdf.Ln(1)
	}

	var buf bytes.Buffer
	if err := pdf.Output(&buf); err != nil {
		return nil, err
	}
	return buf.Bytes(), nil
}

// Observe derives measured, rule-based statements from a result. It reports
// what happened (facts); any mitigation reading is labeled as such.
func Observe(res Input) []string {
	var out []string
	m := res
	total := m.TotalRequests
	fail := res.FailedRequests
	if total < 0 {
		total = 0
	}

	limited := m.StatusCodes["429"]
	fivexx := sumCodes(m.StatusCodes, 500, 599)
	fourxx := sumCodes(m.StatusCodes, 400, 499) - limited

	defense := "off"
	if res.Defense {
		defense = "on"
	}

	if res.Defense {
		if limited > 0 {
			out = append(out, fmt.Sprintf(
				"Defense was on: the rate limiter rejected %d excess request(s) with HTTP 429 (%.1f%% of total).",
				limited, pct(limited, total)))
		} else {
			out = append(out, "Defense was on but never engaged — traffic stayed under the rate limit (no HTTP 429).")
		}
	} else {
		out = append(out, fmt.Sprintf("Defense was %s: this run is an unprotected baseline.", defense))
	}

	if fivexx > 0 {
		out = append(out, fmt.Sprintf("HTTP 5xx responses appeared during the experiment (%d).", fivexx))
	}
	if fourxx > 0 {
		out = append(out, fmt.Sprintf("HTTP 4xx (non-429) responses appeared (%d) — possible overload or client errors.", fourxx))
	}
	if fail == 0 {
		out = append(out, "No failed requests recorded.")
	} else if !res.Defense || limited < fail {
		out = append(out, fmt.Sprintf("Error rate was %.1f%% (%d of %d requests).", pct(fail, total), fail, total))
	}

	if m.P50LatencyMs > 0 && m.P95LatencyMs >= 3*m.P50LatencyMs {
		out = append(out, fmt.Sprintf(
			"Tail latency inflated: P95 (%.1f ms) was well above P50 (%.1f ms) under load.",
			m.P95LatencyMs, m.P50LatencyMs))
	} else if m.P50LatencyMs > 0 {
		out = append(out, fmt.Sprintf(
			"Latency stayed even: P50 %.1f ms, P95 %.1f ms.",
			m.P50LatencyMs, m.P95LatencyMs))
	}
	if total == 0 {
		out = append(out, "No requests completed — check that the test server was reachable.")
	}
	return out
}

func configRows(res Input) [][2]string {
	c := res
	defense := "off — unprotected baseline"
	if c.Defense {
		defense = "on — rate limiting"
	}
	return [][2]string{
		{"Endpoint", res.TargetURL},
		{"Duration", fmt.Sprintf("%d s", c.DurationSeconds)},
		{"Target rate", fmt.Sprintf("%d req/s", c.RequestsPerSecond)},
		{"Workers", fmt.Sprintf("%d", c.Workers)},
		{"Defense", defense},
		{"Final status", res.Status},
	}
}

func trafficRows(res Input) [][2]string {
	m := res
	total := m.TotalRequests
	elapsed := elapsedSeconds(res)
	avgRPS := 0.0
	if elapsed > 0 {
		avgRPS = float64(total) / elapsed
	}
	return [][2]string{
		{"Total requests", fmt.Sprintf("%d", total)},
		{"Successful", fmt.Sprintf("%d", res.SuccessfulRequests)},
		{"Failed", fmt.Sprintf("%d", res.FailedRequests)},
		{"Error rate", fmt.Sprintf("%.1f%%", pct(res.FailedRequests, total))},
		{"Average rate", fmt.Sprintf("%.1f req/s (total / elapsed)", avgRPS)},
		{"Rate at completion", fmt.Sprintf("%.1f req/s (trailing 5 s window)", m.RPS)},
	}
}

func latencyRows(res Input) [][2]string {
	m := res
	return [][2]string{
		{"Average", fmt.Sprintf("%.1f ms", m.AvgLatencyMs)},
		{"P50", fmt.Sprintf("%.1f ms", m.P50LatencyMs)},
		{"P95", fmt.Sprintf("%.1f ms", m.P95LatencyMs)},
		{"P99", fmt.Sprintf("%.1f ms", m.P99LatencyMs)},
		{"Samples", fmt.Sprintf("%d", m.Samples)},
	}
}

func serverRows(res Input) [][2]string {
	rows := [][2]string{}
	if res.CPUPercent != nil {
		rows = append(rows, [2]string{"Target CPU", fmt.Sprintf("%.1f%%", *res.CPUPercent)})
	} else {
		rows = append(rows, [2]string{"Target CPU", "n/a (non-Linux target)"})
	}
	if res.MemoryRSSBytes != nil {
		rows = append(rows, [2]string{"Target memory (RSS)", fmt.Sprintf("%.1f MiB", float64(*res.MemoryRSSBytes)/1048576)})
	} else {
		rows = append(rows, [2]string{"Target memory (RSS)", "n/a (non-Linux target)"})
	}
	codes := res.StatusCodes
	if len(codes) == 0 {
		rows = append(rows, [2]string{"Status codes", "none recorded"})
		return rows
	}
	keys := make([]string, 0, len(codes))
	for k := range codes {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	for _, k := range keys {
		rows = append(rows, [2]string{"HTTP " + k, fmt.Sprintf("%d", codes[k])})
	}
	return rows
}

// --- layout primitives ---

func footer(pdf *fpdf.Fpdf) {
	pdf.SetFooterFunc(func() {
		pdf.SetY(-15)
		pdf.SetFont("Helvetica", "", 8)
		pdf.SetTextColor(130, 130, 130)
		pdf.CellFormat(0, 5,
			fmt.Sprintf("Local lab report — not for production use  ·  page %d", pdf.PageNo()),
			"", 0, "C", false, 0, "")
	})
}

func coverStrip(pdf *fpdf.Fpdf, res Input) {
	pdf.SetFillColor(accentR, accentG, accentB)
	pdf.SetTextColor(255, 255, 255)
	pdf.SetFont("Helvetica", "B", 20)
	pdf.CellFormat(0, 12, "DDoSLab Experiment Report", "", 1, "L", true, 0, "")
	pdf.SetFont("Helvetica", "", 10)
	pdf.CellFormat(0, 7,
		fmt.Sprintf("Experiment %s  ·  %s  ·  %s",
			res.ID, res.StartedAt.Format(time.RFC3339), res.Status),
		"", 1, "L", true, 0, "")
	pdf.Ln(6)
}

func section(pdf *fpdf.Fpdf, title string) {
	pdf.SetFont("Helvetica", "B", 13)
	pdf.SetTextColor(accentR, accentG, accentB)
	pdf.CellFormat(0, 8, title, "", 1, "L", false, 0, "")
	pdf.Ln(1)
}

func pairs(pdf *fpdf.Fpdf, rows [][2]string) {
	for _, r := range rows {
		pdf.SetFont("Helvetica", "", 10)
		pdf.SetTextColor(110, 110, 110)
		pdf.CellFormat(62, 6.5, r[0], "", 0, "L", false, 0, "")
		pdf.SetTextColor(20, 20, 20)
		pdf.CellFormat(0, 6.5, r[1], "", 1, "L", false, 0, "")
	}
	pdf.Ln(2)
}

func rule(pdf *fpdf.Fpdf) {
	x := pdf.GetX()
	pdf.SetDrawColor(220, 220, 220)
	pdf.Line(x, pdf.GetY(), 210-marginMM, pdf.GetY())
	pdf.Ln(4)
}

func elapsedSeconds(res Input) float64 {
	end := time.Now().UTC()
	if res.CompletedAt != nil {
		end = *res.CompletedAt
	}
	return end.Sub(res.StartedAt).Seconds()
}

func pct(part, total int64) float64 {
	if total <= 0 {
		return 0
	}
	return float64(part) / float64(total) * 100
}

func sumCodes(codes map[string]int64, from, to int) int64 {
	var n int64
	for k, v := range codes {
		var c int
		if _, err := fmt.Sscanf(k, "%d", &c); err == nil && c >= from && c <= to {
			n += v
		}
	}
	return n
}
